File: mpp-proiect-java-culbec/MotorcycleContest/Server/src/main/java/Service/ServiceParticipantImpl.java
package Service;

import Exceptions.ServiceException;
import Model.Participant;
import Repository.ParticipantRepository;
import Repository.RepositoryException;
import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;

import java.util.List;
import java.util.Optional;
import java.util.UUID;

public class ServiceParticipantImpl implements ServiceParticipant {
    private final Logger logger = LogManager.getLogger(this.getClass());
    private final ParticipantRepository participantRepository;

    public ServiceParticipantImpl(ParticipantRepository participantRepository) {
        logger.traceEntry("Initializing the ServiceParticipantImpl with the ParticipantRepository...");
        this.participantRepository = participantRepository;
        logger.traceExit("Initialized the ServiceParticipantImpl with the ParticipantRepository!");
    }


    @Override
    public Iterable<Participant> findParticipantsByTeam(String team) throws ServiceException {
        logger.traceEntry("Finding all participants from the team {}...", team);

        try {
            List<Participant> participants = (List<Participant>) participantRepository.findParticipantsByTeam(team);

            if (participants.isEmpty()) {
                logger.error("The team {} doesn't exist!", team);
                throw new ServiceException("The team doesn't exist!");
            }

            logger.traceExit("Found all participants from the team {}!", team);
            return participants;
        } catch (RepositoryException e) {
            logger.error("Couldn't find the participants from the team {}: {}", team, e.getMessage());
            throw new ServiceException(e.getMessage());
        }
    }

    @Override
    public Participant addParticipant(String firstName, String lastName, String team, int engineCapacity) throws ServiceException {
        logger.traceEntry("Adding the participant with the first name {}, last name {}, team {} and engine capacity {}...", firstName, lastName, team, engineCapacity);

        logger.info("Finding the participant by the fields {}, {}, {}, {}", firstName, lastName, team, engineCapacity);
        Participant participant = new Participant.Builder()
                .setId(UUID.randomUUID())
                .setFirstName(firstName)
                .setLastName(lastName)
                .setTeam(team)
                .setEngineCapacity(engineCapacity)
                .build();

        try {
            Optional<Participant> saved = this.participantRepository.save(participant);

            if (saved.isEmpty()) {
                logger.error("The participant with the first name {}, last name {}, team {} and engine capacity {} already exists!", firstName, lastName, team, engineCapacity);
                throw new ServiceException("The participant already exists!");
            }

            logger.traceExit("The participant was saved. Returning...");
            return saved.get();
        } catch (RepositoryException e) {
            logger.error("Something wrong happened when trying to find the participant by the fields: {}", e.getMessage());
            throw new ServiceException("Something wrong happened when trying to find the participant by the fields:" + e.getMessage());
        }
    }
}

File: mpp-proiect-csharp-culbec/Server/Service/ServiceParticipantImpl.cs
using Common.Exceptions;
using log4net;
using Model;
using Persistence.Repository;

namespace Server.Service;

public class ServiceParticipantImpl : IServiceParticipant
{
    private static readonly ILog Log = LogManager.GetLogger(typeof(ServiceParticipantImpl));
    private readonly IRepositoryParticipant _repositoryParticipant;

    public ServiceParticipantImpl(IRepositoryParticipant repositoryParticipant)
    {
        Log.Info("Initializing ServiceParticipantImpl.");
        _repositoryParticipant = repositoryParticipant;
        Log.Info("ServiceParticipantImpl initialized.");
    }

    public IEnumerable<Participant> FindParticipantsByTeam(string team)
    {
        Log.InfoFormat("Searching for participants of team {0}.", team);
        try
        {
            var participants = _repositoryParticipant.FindParticipantsByTeam(team).ToList();

            Log.InfoFormat("Found {0} participants of team {1}.", participants.Count, team);
            return participants;
        }
        catch (RepositoryException e)
        {
            Log.ErrorFormat("Something went wrong when searching for the participants of a team: {0}", e.Message);
            throw new ServiceException("Something went wrong when searching for the participants of a team: " +
                                       e.Message);
        }
    }

    public Participant AddParticipant(string firstName, string lastName, string team, int engineCapacity)
    {
        Log.InfoFormat("Adding participant {0} {1} to team {2} with engine capacity {3}.", firstName, lastName, team,
            engineCapacity);

        try
        {
            Log.Info("Creating the participant...");
            var participant = Participant.NewBuilder()
                .SetId(Guid.NewGuid())
                .SetFirstName(firstName)
                .SetLastName(lastName)
                .SetTeam(team)
                .SetEngineCapacity(engineCapacity)
                .Build();
            Log.Info("Participant created.");

            Log.Info("Attempting to add the participant to the repository...");
            var result = _repositoryParticipant.Save(participant);

            if (result != null)
            {
                Log.Error("The participant already exists!");
                throw new ServiceException("The participant already exists!");
            }

            Log.Info("Participant added successfully.");
            return participant;
        }
        catch (RepositoryException e)
        {
            Log.ErrorFormat("Something went wrong when trying to add a participant: {0}", e.Message);
            throw new ServiceException("Something went wrong when trying to add a participant: " + e.Message);
        }
    }
}